##### Britt van Gemert  
###### s4555740  
##### Robert Frowein  
###### s4425960  

# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [2]:
import Toolbox.apriorimining as am



#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `associationmining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [3]:
am.generate_association_rules()


Please enter support value in %: 80
Please enter confidence value in %: 100
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename and extension: Data/courses.txt
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 100.0
set= { 2 },  sup= 83.33
set= { 7 },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 83.0
set= { 6, 7 },  sup= 83.0
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00
--------------------------------------------------------


Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00

The conclusions you can make based on these association rules are:  
*Rule #1*: 100% of all students follow physics.  
*Rule #2*: 83.3% of all students that follow course 2 also follow course 6.  
*Rule #3*: 83.3% of all students that follow course 7 also follow course 6.


  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? You can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [8]:
am.generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 80
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename and extension: Data/MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1983), Star Wars (1977) },  sup= 50.0
set= { Fargo (1

Associations with the strongest confidence (we choose the highest 4):  
Rule #73: { 172, 174, 181 } ==> { 50 }, sup= 31.18, conf= 99.66
  
Rule #49: { 172, 181 } ==> { 50 }, sup= 33.40, conf= 99.37
  
Rule #58: { 181, 56 } ==> { 50 }, sup= 30.22, conf= 98.62
  
Rule #37: { 174, 181 } ==> { 50 }, sup= 35.74, conf= 98.54
  
172 = The Empire Strikes Back (1980)  
174 = Raiders of the Lost Ark (1981)  
181 = Return of the Jedi (1983)  
50 = Star Wars (1977)  
56 = Pulp Fiction (1994)  
  
Yes, the associatioins make sense, because a person that has already watched the sequels (and Indiana Jones), will probably also have watched the prequel(Star Wars).

#### 5.2.3 
(1 point) Which movie has been bought by the most users? There are only few rules with more than three items. Why?

In [ ]:
am.generate_association_rules()

Please enter support value in %: 10
Please enter confidence value in %: 0
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename and extension: Data/MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1983), Star Wars (1977) },  sup= 50.0
set= { Fargo (19

set= { Star Wars (1977) },  sup= 61.82  
Star Wars (50) has been bought by the most users.  
There are only a few rules with more than three items because it is hard to associate (with confidence) so many different movies together to one movie. This is because of the numbers we chose for our association rules, say if we aimed for a lower confidence, then we would have more cases with more than three items.

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

Yes, this is possible. For example:  
Say a movie only appears three times but always with the same other movie, then the confidence would be 100%. However, the support would only be 3. This is because support is only a measurement of how many times a certain item occurs in the data set.

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

4. (1.5 points) Prove that if the confidence of the rule $ \text{ {a } } \rightarrow   \text{{b }} $ is less than the support of $ \text{ {b }}$  then
$$
c(  \text{ {~a } } \rightarrow   \text{{b }})> c(  \text{ {a } } \rightarrow   \text{{b }})
$$
and
$$
c(  \text{ {~a } } \rightarrow   \text{{b }})> s( {b })
$$

where $ \text{c(.)}$  denotes the rule confidence and  $ \text{s(.)}$ denote the support of an itemset. 

Hint: To prove the statement rewrite the confidence and the support of the rule in terms of probabilities: 
$$
c(  \text{ {a } } \rightarrow   \text{{b }})=  \frac{P( {a,b })}{P( {a })}
$$

and

$$
s( {b })=P( {b })
$$

Further more, make use of probability rules such as $p(b) = p(\text{~}a,b) + p(a,b)$ and $p(\text{~}a) = 1 - p(a)$ and algebraic rules such as $$ \frac{p(a,b)}{p(b)} < p(a) \rightarrow 1 - \frac{p(a,b)}{p(b)} > 1 - p(a) $$


1. c = $\frac{(a,b)}{(a)} = \frac{30\%}{45\%} = \frac {0.30}{0.45} \approx $ 0.667  
This is 66,7%, so yes this is interesting because it is above the 60% treshold of the confidence.  
2. Lift = $\frac {P(B|A)}{P(B)} = \frac {0.667}{0.80} \approx $ 0.83   
0.83 is < 1, so a and b are negatively associated. The items are thus not highly dependent.  
3.  
From exercise 1 we learn that the two items together have a high (enough) confidence. However, from 2 we
learn that they are also negatively associated. You would think that these things contradict each other, which
make the outcomes a bit weird. Nevertheless, this outcome is correct because, even though the two items
are not frequently bought together (negatively associated), they occur so many times together in the data
set that the confidence is still high enough to exceed the threshold.  
4. If $\frac{p(a,b)}{p(a)}$ < p(b)  
1) Then $\frac{p(\text{~}a,b)}{p(\text{~}a)}$ > $\frac{p(a,b)}{p(a)}$
  
> __Proof__:
  
> $\frac{p(\text{~}a,b)}{p(\text{~}a)}$ > $\frac{p(a,b)}{p(a)}$  
  
> $\frac{p(\text{~}a,b)}{p(\text{~}a)}$ > p(b|a) $\hspace{10mm}$ *(Bayes rule)* 
  
> p($\text{~}$a,b) > p(b|a) $*$ p(~a)  
  
> p($\text{~}$a,b) + p(a,b) > p(b|a) $*$ p($\text{~}$a) + p(a,b) 
  
> p(b) > p(b|a) $*$ p($\text{~}$a) + p(a,b)  
  
> p(b) > $\frac{p(a,b) * p(\text{~}a)}{p(a)}$ + p(a,b)  
  
> p(b) > $\frac{p(a,b) * (1 - p(a))}{p(a)}$ + p(a,b)  
  
> p(b) > $\frac{p(a,b) - p(a,b) * p(a)}{p(a)}$ + p(a,b)  
  
> p(b) > $\frac{p(a,b)}{p(a)} - \frac{p(a,b) * p(a)}{p(a)}$ + p(a,b)  
  
> p(b) > $\frac{p(a,b)}{p(a)}$ - p(a,b) + p(a,b)  
  
> p(b) > $\frac{p(a,b)}{p(a)}$  
  
> $\frac{p(a,b)}{p(a)}$ < p(b)  
  
> This is correct, because it is exactly the same as the if-statement.


2) also $\frac{p(\text{~}a,b)}{p(\text{~}a)}$ > p(b)
  
> __Proof__:
  
> $\frac{p(\text{~}a,b)}{p(\text{~}a)}$ > p(b) 
  
> $\frac{p(\text{~}a,b)}{p(\text{~}a)}$ > p($\text{~}$a,b) + p(a,b)  
  
> p($\text{~}$a,b) > (p($\text{~}$a,b) + p(a,b)) $*$ p($\text{~}$a)  
  
> p($\text{~}$a,b) > (p($\text{~}$a,b) + p(a,b)) $*$ (1 - p(a))  
  
> p($\text{~}$a,b) > p($\text{~}$a,b) - (p($\text{~}$a,b) $*$ p(a)) + p(a,b) - (p(a,b) $*$ p(a))  
  
> $\frac{p(\text{~}a,b)}{p(a)}$ > $\frac{p(\text{~}a,b)}{p(a)}$ - p($\text{~}$a,b) + $\frac{p(a,b)}{p(a)}$ - p(a,b)  
  
> $\frac{p(\text{~}a,b)}{p(a)}$ > $\frac{p(\text{~}a,b)}{p(a)}$ + $\frac{p(a,b)}{p(a)}$ - p(b)  
  
> 0 > $\frac{p(a,b)}{p(a)}$ - p(b)  
  
> This is correct, because in the if-statement p(b) is larger than $\frac{p(a,b)}{p(a)}$, which makes this correct


#### 5.3.2

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



Suppose A = high-definition and B = exercise machine  
**1:**  
  
*Table 2*  
Odd ratio = $\frac {\frac{105}{294} \hspace{2mm}\cdot \hspace{2mm} \frac{62}{294}}{\frac{87}{294} \hspace{2mm}\cdot\hspace{2mm} \frac{40}{294}} \approx$ 1.871  
  
*Table 3 College students*  
Odd ratio = $\frac {\frac{2}{36} \hspace{2mm}\cdot\hspace{2mm} \frac{20}{36}}{\frac{9}{36} \hspace{2mm}\cdot\hspace{2mm} \frac{5}{36}} \approx$ 0.889 
  
*Table 3 Working students*  
Odd ratio = $\frac {\frac{103}{258} \hspace{2mm}\cdot\hspace{2mm} \frac{42}{258}}{\frac{78}{258} \hspace{2mm}\cdot\hspace{2mm} \frac{35}{258}} \approx$ 1.585    
  
The direction of the odd ratio changes when the data is pooled together such that the ratio is higher than
the two separate groups individually. This would mean that in an overall population, people are more likely
to have both an HDTV and an Exercise machine, just like Adults also are more likely to have them both in
their presence, however, College students are less likely to have them both. Nonetheless, you can see that the
pooled together group is more in accordance with Adults than with College students.  
  
**2:**  
  
*Table 2*  
$\Phi-coefficient = \frac {\frac{105}{294} - \frac{192}{294} \hspace{2mm}\cdot\hspace{2mm} \frac{145}{294}}{\sqrt {\frac{192}{294} \hspace{2mm}\cdot\hspace{2mm} (1-\frac{192}{294}) \hspace{2mm}\cdot\hspace{2mm} \frac{145}{294} \hspace{2mm}\cdot\hspace{2mm} (1-\frac{145}{294})}} \approx$ 0.147  
  
*Table 3 College students*  
$\Phi-coefficient = \frac {\frac{2}{36} - \frac{11}{36} \hspace{2mm}\cdot\hspace{2mm} \frac{7}{36}}{\sqrt {\frac{11}{36} \hspace{2mm}\cdot\hspace{2mm} (1-\frac{11}{36}) \hspace{2mm}\cdot\hspace{2mm} \frac{7}{36} \hspace{2mm}\cdot\hspace{2mm} (1-\frac{7}{36})}} \approx$ -0.021  
  
*Table 3 Working students*  
$\Phi-coefficient = \frac {\frac{103}{258} - \frac{181}{258} \hspace{2mm}\cdot\hspace{2mm} \frac{138}{258}}{\sqrt {\frac{181}{258} \hspace{2mm}\cdot\hspace{2mm} (1-\frac{181}{258}) \hspace{2mm}\cdot\hspace{2mm} \frac{138}{258} \hspace{2mm}\cdot\hspace{2mm} (1-\frac{138}{258})}} \approx$ 0.105  
  
The $\phi-coefficient$ of Table 2 is above the coefficients
of College students and Adults separately. Nevertheless,
all these computed coefficients tell us that there is no or little correlation between the items.  
  
**3:**  
  
*Table 2*  
$Interest = \frac {\frac {105}{294}}{\frac {192}{294} \hspace{2mm}\cdot\hspace{2mm} \frac {145}{294}} \approx$ 1.109  
  
*Table 3 College students*  
$Interest = \frac {\frac {2}{36}}{\frac {11}{36} \hspace{2mm}\cdot\hspace{2mm} \frac {7}{36}} \approx$ 0.935    
  
*Table 3 Working students*  
$Interest = \frac {\frac {103}{258}}{\frac {181}{258} \hspace{2mm}\cdot\hspace{2mm} \frac {138}{258}} \approx$ 1.064   
  
The interest of Table 2 is > 1, which means that the items are positively associated. However, College students are
negatively associated (0.935 < 1), but then again Adults are also positively associated. This means that the
pooled together group is more in accordance with Adults yet again.